### PR TITLE
Allow TLS when using Sentinel

### PR DESCRIPTION
**Pull request by fenichelar**
Fixes https://github.com/luin/ioredis/issues/740

### CODE REVIEW

**Comment by fenichelar** [comment]
Tested with Redis Enterprise and is working as expected. Redis Enterprise requires client certificates. Code used for testing:

```js
const fs = require('fs');
const ioredis = require('ioredis');

const redisClient = new ioredis({
  sentinels: [
    {
      host: '10.0.1.1',
      port: 8001,
    },
    {
      host: '10.0.1.2',
      port: 8001,
    },
    {
      host: '10.0.1.3',
      port: 8001,
    },
  ],
  name: 'test',
  tls: {
    key: fs.readFileSync('key.pem').toString(),
    cert: fs.readFileSync('cert.pem').toString(),
    ca: fs.readFileSync('ca.pem').toString(),
    secureProtocol: 'TLSv1_2_method',
    checkServerIdentity: () => {
      return undefined;
    },
  },
});

redisClient.on('error', err => {
  console.error(err);
});

redisClient.on('ready', () => {
  console.log('ready');
  process.exit(0);
});
```

**Comment by luin** [comment]
Thanks for the pull request! A question here, does connecting to sentinel servers need tls support?

**Comment by fenichelar** [comment]
@luin I don't think so. In Redis Enterprise, you enable TLS per database. However, only a single Sentinel runs per node. If you have 2 database on a single node, one can require TLS and the other can not (which is what I was using for testing actually). I don't think there is an option to require TLS for the connection to Sentinel, but I'll look at the documentation again. That would be nice though because a MITM attack against Sentinel could be quite bad.

**Comment by fenichelar** [comment]
Also, I should clarify, for Redis Enterprise it is not actually Sentinel, it is Discovery Service. The Discovery Service API is intentionally compliant with Sentinel but it isn't actually the same thing. See https://docs.redislabs.com/latest/rs/concepts/data-access/discovery-service/.

**Comment by fenichelar** [comment]
It worked! While it doesn't appear to be possible to require TLS connections to the Discovery Service, it appears to be possible to use TLS to connect.

Should I modify this PR to encrypt the connection to Sentinel / Discovery Service if TLS options are supplied or make a seperate PR?

**Comment by fenichelar** [comment]
It was so easy to implement I just added it here.

Please let me know if there is anything else.

File: lib/connectors/SentinelConnector/index.ts [hunk 1]
```diff
@@ -104,7 +104,12 @@ export default class SentinelConnector extends AbstractConnector {
         }
         if (resolved) {
           debug('resolved: %s:%s', resolved.host, resolved.port)
-          _this.stream = createConnection(resolved)
+          if (_this.options.tls) {
```
Comment (luin) [inline lib/connectors/SentinelConnector/index.ts]: if (_this.options.enableTLSForSentinel && _this.options.tls) {

File: lib/connectors/SentinelConnector/index.ts [hunk 2]
```diff
@@ -193,6 +198,7 @@ export default class SentinelConnector extends AbstractConnector {
       port: endpoint.port || 26379,
       host: endpoint.host,
       family: endpoint.family || (isIIpcConnectionOptions(this.options) ? undefined : this.options.family),
+      tls: this.options.tls,
```
Comment (luin) [inline lib/connectors/SentinelConnector/index.ts]: tls: this.options.enableTLSForSentinel && this.options.tls

**Comment by fenichelar** [comment]
This makes sense. However, what if someone wanted just the connection to Redis to be encrypted, not Sentinel? Or the other way around?

What if we add `enableTLSForSentinel` and `enableTLSForRedis`? Both will default to `false` for this version. In v5, switch them both to default to `true`. Thoughts?

**Comment by luin** [comment]
We can provide two options in this version:
1. `enableTLSForSentinelMode` (default value is `false`). If true, the `tls` option will be used to connect to the redis server that resolved from sentinels.
2. `sentinelTLS` (default value is `undefined`). Similar to the `tls` option but used for connecting to sentinels.

In v5, `enableTLSForSentinelMode` will be removed and `sentinelTLS` will remain intact. I've changed the comments I left below the code accordingly. What do you think?

**Comment by fenichelar** [comment]
Makes sense. How's that look?

File: lib/connectors/SentinelConnector/index.ts [hunk 3]
```diff
@@ -104,7 +106,12 @@ export default class SentinelConnector extends AbstractConnector {
         }
         if (resolved) {
           debug('resolved: %s:%s', resolved.host, resolved.port)
-          _this.stream = createConnection(resolved)
+          if (_this.options.enableTLSForSentinelMode && _this.options.tls) {
+            Object.assign(resolved, _this.options.tls)
```
Comment (luin) [inline lib/connectors/SentinelConnector/index.ts]: Shouldn't `Object.assign(resolved, _this.options.tls)` be `Object.assign(resolved, {tls: _this.options.tls})`?

Comment (fenichelar) [reply lib/connectors/SentinelConnector/index.ts]: Hahaha, that's what I thought too and I spent so much time trying to figure out why it wasn't working!!

https://nodejs.org/api/tls.html#tls_tls_connect_options_callback

> secureContext: TLS context object created with tls.createSecureContext(). If a secureContext is not provided, one will be created by passing the entire options object to tls.createSecureContext().


The entire options object is passed to `createSecureContext()`.

Comment (fenichelar) [reply lib/connectors/SentinelConnector/index.ts]: The `StandardConnector` works the same: https://github.com/luin/ioredis/blob/master/lib/connectors/StandaloneConnector.ts#L45-L47

**Comment by fenichelar** [comment]
@luin Awesome. I'm trying to setup a new Redis cluster this week with TLS. When do you think a new version of ioredis could be released with this change? Thanks.

**Comment by luin** [comment]
The new version has been released automatically: v4.5.0. The version is released under the next tag (can be installed via npm i `ioredis@next` or `npm i ioredis@4.5.0`) and will be released to the latest tag in a month.

**Comment by fenichelar** [comment]
Oh, awesome!! Thanks :)

**Pull request merged**